src/circuit/gadget/ecc/chip/mul.rs

```diff
@@ -0,0 +1,472 @@
+}
+
+impl<C: CurveAffine> From<&EccConfig<C>> for Config<C> {
+    fn from(ecc_config: &EccConfig<C>) -> Self {
+        let config = Self {
+            q_mul_complete: ecc_config.q_mul_complete,
+            q_mul_decompose_var: ecc_config.q_mul_decompose_var,
+            q_init_z: ecc_config.q_init_z,
+            z_complete: ecc_config.advices[9],
+            scalar: ecc_config.advices[1],
+            perm: ecc_config.perm.clone(),
+            add_config: ecc_config.into(),
+            hi_config: incomplete::Config::hi_config(ecc_config),
+            lo_config: incomplete::Config::lo_config(ecc_config),
+        };
+
+        assert_eq!(
+            config.hi_config.x_p, config.lo_config.x_p,
+            "x_p is shared across hi and lo halves."
+        );
+        assert_eq!(
+            config.hi_config.y_p, config.lo_config.y_p,
+            "y_p is shared across hi and lo halves."
+        );
+
+        let add_config_advices = config.add_config.advice_columns();
+        assert!(
+            !add_config_advices.contains(&config.z_complete),
+            "z_complete cannot overlap with complete addition columns."
+        );
+        assert!(
+            !add_config_advices.contains(&config.hi_config.z),
+            "hi_config z cannot overlap with complete addition columns."
+        );
+
+        config
+    }
+}
+
+impl<C: CurveAffine> Config<C>
+where
+    C::Scalar: PrimeFieldBits,
+{
+    pub(super) fn create_gate(&self, meta: &mut ConstraintSystem<C::Base>) {
+        self.hi_config.create_gate(meta);
+        self.lo_config.create_gate(meta);
+
+        let complete_config: complete::Config<C> = self.into();
+        complete_config.create_gate(meta);
+
+        self.create_init_scalar_gate(meta);
+        self.create_final_scalar_gate(meta);
+    }
+
+    /// Gate used to check that the running sum for scalar decomposition is initialized to zero.
+    fn create_init_scalar_gate(&self, meta: &mut ConstraintSystem<C::Base>) {
+        meta.create_gate("Initialize running sum for variable-base mul", |meta| {
+            let q_init_z = meta.query_selector(self.q_init_z);
+            let z = meta.query_advice(self.hi_config.z, Rotation::cur());
+
+            vec![q_init_z * z]
+        });
+    }
+
+    /// Gate used to check final scalar is recovered.
+    fn create_final_scalar_gate(&self, meta: &mut ConstraintSystem<C::Base>) {
+        meta.create_gate("Decompose scalar for variable-base mul", |meta| {
+            let q_mul_decompose_var = meta.query_selector(self.q_mul_decompose_var);
+            let scalar = meta.query_advice(self.scalar, Rotation::cur());
+            let z_cur = meta.query_advice(self.z_complete, Rotation::cur());
+
+            // The scalar field `F_q = 2^254 + t_q`.
+            // -((2^127)^2) = -(2^254) = t_q (mod q)
+            let t_q = -(C::Scalar::from_u128(1u128 << 127).square());
+            let t_q = C::Base::from_bytes(&t_q.to_bytes()).unwrap();
```

Comment:
TODO: add `t_q` (and corresponding `t_p` for Pallas) as constants.

_Originally posted by @therealyingtong in https://github.com/zcash/orchard/pull/107#discussion_r646072661_